File: frontend/src/lib/types/alert.ts
export type AlertPriority = 'P1' | 'P2' | 'P3' | 'P4' | 'P5';
export type AlertStatus = 'open' | 'acknowledged' | 'closed' | 'snoozed';
export type AlertSource = 'webhook' | 'api' | 'email' | 'integration' | 'manual';

export interface Alert {
	id: string;
	organization_id: string;
	source: AlertSource;
	source_id?: string;
	priority: AlertPriority;
	status: AlertStatus;
	message: string;
	description?: string;
	tags: string[];
	custom_fields: Record<string, unknown>;

	// Assignment
	assigned_to_user_id?: string;
	assigned_to_team_id?: string;

	// Acknowledgment
	acknowledged_by?: string;
	acknowledged_at?: string;

	// Closure
	closed_by?: string;
	closed_at?: string;
	close_reason?: string;

	// Snooze
	snoozed_until?: string;

	// Escalation
	escalation_policy_id?: string;
	escalation_level: number;
	last_escalated_at?: string;

	created_at: string;
	updated_at: string;
}

export interface CreateAlertRequest {
	source: string;
	source_id?: string;
	priority: AlertPriority;
	message: string;
	description?: string;
	tags?: string[];
	custom_fields?: Record<string, unknown>;
}

export interface UpdateAlertRequest {
	priority?: AlertPriority;
	message?: string;
	description?: string;
	tags?: string[];
	custom_fields?: Record<string, unknown>;
}

export interface CloseAlertRequest {
	reason: string;
}

export interface SnoozeAlertRequest {
	until: string; // ISO 8601 date string
}

export interface AssignAlertRequest {
	user_id?: string;
	team_id?: string;
}

export interface ListAlertsParams {
	status?: AlertStatus[];
	priority?: AlertPriority[];
	assigned_to_user?: string;
	assigned_to_team?: string;
	source?: string;
	search?: string;
	page?: number;
	page_size?: number;
}

export interface ListAlertsResponse {
	alerts: Alert[];
	total: number;
	page: number;
	page_size: number;
}

File: backend/internal/domain/alert.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

type Alert struct {
	ID             uuid.UUID
	OrganizationID uuid.UUID
	Source         string
	SourceID       *string
	Priority       AlertPriority
	Status         AlertStatus
	Message        string
	Description    *string
	Tags           []string
	CustomFields   map[string]interface{}

	// Assignment
	AssignedToUserID *uuid.UUID
	AssignedToTeamID *uuid.UUID

	// Acknowledgment
	AcknowledgedBy *uuid.UUID
	AcknowledgedAt *time.Time

	// Closure
	ClosedBy    *uuid.UUID
	ClosedAt    *time.Time
	CloseReason *string

	// Snooze
	SnoozedUntil *time.Time

	// Escalation
	EscalationPolicyID *uuid.UUID
	EscalationLevel    int
	LastEscalatedAt    *time.Time

	// Deduplication
	DedupKey          *string
	DedupCount        int
	FirstOccurrenceAt *time.Time
	LastOccurrenceAt  *time.Time

	CreatedAt time.Time
	UpdatedAt time.Time
}

type AlertPriority string

const (
	PriorityP1 AlertPriority = "P1" // Critical
	PriorityP2 AlertPriority = "P2" // High
	PriorityP3 AlertPriority = "P3" // Medium
	PriorityP4 AlertPriority = "P4" // Low
	PriorityP5 AlertPriority = "P5" // Informational
)

func (p AlertPriority) String() string {
	return string(p)
}

func (p AlertPriority) IsValid() bool {
	switch p {
	case PriorityP1, PriorityP2, PriorityP3, PriorityP4, PriorityP5:
		return true
	}
	return false
}

type AlertStatus string

const (
	AlertStatusOpen         AlertStatus = "open"
	AlertStatusAcknowledged AlertStatus = "acknowledged"
	AlertStatusClosed       AlertStatus = "closed"
	AlertStatusSnoozed      AlertStatus = "snoozed"
)

func (s AlertStatus) String() string {
	return string(s)
}

func (s AlertStatus) IsValid() bool {
	switch s {
	case AlertStatusOpen, AlertStatusAcknowledged, AlertStatusClosed, AlertStatusSnoozed:
		return true
	}
	return false
}

type AlertSource string

const (
	AlertSourceWebhook     AlertSource = "webhook"
	AlertSourceAPI         AlertSource = "api"
	AlertSourceEmail       AlertSource = "email"
	AlertSourceIntegration AlertSource = "integration"
	AlertSourceManual      AlertSource = "manual"
)

func (s AlertSource) String() string {
	return string(s)
}

// AlertWithAssignees includes assigned user and team information
type AlertWithAssignees struct {
	Alert
	AssignedUser *User
	AssignedTeam *Team
}

// AlertFilter for filtering and pagination
type AlertFilter struct {
	OrganizationID uuid.UUID
	Status         []AlertStatus
	Priority       []AlertPriority
	AssignedToUser *uuid.UUID
	AssignedToTeam *uuid.UUID
	Source         *string
	Search         *string // Search in message and description
	Limit          int
	Offset         int
}
